Update Scalar DB backup and restoration guide (#486)
scalar-labs/scalardb: docs/backup-restore.md

With [scalar-admin client tool](https://github.com/scalar-labs/scalar-admin/tree/scalar-admin-dockerfile#client-side-tool), you can pause nodes/servers/applications that implement the scalar-admin interface without losing ongoing transactions.

Note that when you use a point-in-time-restore/recovery mechanism, it is recommended to minimize the clock drifts between clients and servers by using clock synchronization such as NTP.
Otherwise, the time you get as a paused duration might be too different from the time in which the pause was actually conducted, which could restore to a point where ongoing transactions exist.
Also, it is recommended to pause a long enough time (e.g., 10 seconds) and use the mid-time of the paused duration as a restore point since clock synchronization cannot perfectly synchronize clocks between nodes.

#### Database-specific ways to create a transactionally-consistent backup   

**Cassandra**

Cassandra has a built-in replication mechanism, so you do not always have to create a transactionally-consistent backup.

> 

Reviewer: ```suggestion
```

Reviewer 2: Thank you for the suggestion.
Removed blank line.

> For example, if replication is properly set to 3 and only the data of one of the nodes in a cluster is lost, you do not need a transactionally-consistent backup because the node can be recovered with a normal (transactionally-inconsistent) snapshot and the repair mechanism.

Reviewer: ```suggestion
For example, if the replication factor is set to 3 and only the data of one of the nodes in a cluster is lost, you do not need a transactionally-consistent backup because the node can be recovered with a normal (transactionally-inconsistent) snapshot and the repair mechanism.
```

Reviewer 2: Thank you for the suggestion.
Updated document as per suggestion.

However, if the quorum of nodes of a cluster loses their data, we need a transactionally-consistent backup to restore the cluster to a certain transactionally-consistent point.

> If you want to create a transactionally-consistent cluster-wide backup, pause the application and take the snapshots of nodes as described in [the basic strategy](#basic-strategy-to-create-a-transactionally-consistent-backup), or 

Reviewer: ```suggestion
If you want to create a transactionally-consistent cluster-wide backup, pause the Scalar DB application (or Scalar DB server) and take the snapshots of nodes as described in [the basic strategy](#basic-strategy-to-create-a-transactionally-consistent-backup), or 
```

Reviewer 2: Thank you for the suggestion.
Updated document as per suggestion.

stop the Cassandra cluster and take the copies of all the nodes' data, and start the cluster.

To avoid mistakes, it is recommended to use [Cassy](https://github.com/scalar-labs/cassy).

> Cassy is also integrated with `scalar-admin` so it can issue a pause request to the Scalar DB application of a Cassandra cluster.

Reviewer: ```suggestion
Cassy is also integrated with `scalar-admin` so it can issue a pause request to the Scalar DB application (or Scalar DB server) of a Cassandra cluster.
```

Reviewer 2: Thank you for the suggestion.
Updated document as per suggestion.

You need to restore backups appropriately depending on how the backups are created.
For example, if you use MySQL and `mysqldump` to create a backup file, use `mysql` command to restore the file as specified [in the MySQL doc](https://dev.mysql.com/doc/mysql-backup-excerpt/8.0/en/reloading-sql-format-dumps.html). If you use PostgreSQL and `pg_dump` to create a backup file, use `psql` command to restore the file as specified in [the PostgreSQL doc](https://www.postgresql.org/docs/current/backup-dump.html#BACKUP-DUMP-RESTORE).
If you use Amazon RDS (Relational Database Service) or Azure Database for MySQL/PostgreSQL,
you can restore to any point within the backup retention period with the automated backup feature.

### Cassandra

You first need to stop all the nodes of a Cassandra cluster. Clean the directories (`data`, `commitlogs`, and `hints`) and place backups (snapshots) in each node. Then, start all the nodes.

To avoid mistakes, it is recommended to use [Cassy](https://github.com/scalar-labs/cassy).
Please see [the doc](https://github.com/scalar-labs/cassy/blob/master/docs/getting-started.md#take-cluster-wide-consistent-backups) for more details.

### Cosmos DB

> You can follow the [azure official guide](https://docs.microsoft.com/en-us/azure/cosmos-db/restore-account-continuous-backup#restore-account-portal) and change the default consistency to `STRONG` after restoring the backup.

Reviewer: ```suggestion
You can follow the [azure official guide](https://docs.microsoft.com/en-us/azure/cosmos-db/restore-account-continuous-backup#restore-account-portal). After restoring backups. change the default consistencies of the restored databases to `STRONG` 
```

Reviewer 2: Thank you for the suggestion.
Updated document as per suggestion.

You can basically follow [the official doc](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/PointInTimeRecovery.Tutorial.html). However, a table can only be restored with an alias, so you need to restore a table with an alias and delete the original table and rename the alias to the original name to restore the same named tables.

Specifically, please follow the steps below.

1. Create a backup
   * Select the mid-time of paused duration as the restore point.
   * Restore with PITR of table A to another table B
   * Take a backup of the restored table B (assume the backup is named backup B)
   * Remove table B 
2. Restore from the backup 
   * Remove table A
   * Create a table named A with backup B

> Not that tables can only be restored one by one so you need to do the above steps for each table.

Reviewer: ```suggestion
Note that tables can only be restored one by one so you need to do the above steps for each table.
```
